### PR TITLE
feat(api3): add api3 adapter

### PR DESCRIPTION
**Pull request by wiasliaw**
If there are any changes needed, please let me know.

reference
- https://docs.api3.org/guides/dapis/read-a-dapi/

### CODE REVIEW

**Comment by totomanov** [comment]
Hi, thanks for contributing. Could you retarget the PR to the [`experiments`](https://github.com/euler-xyz/euler-price-oracle/tree/experiments) branch instead of `master`?

File: src/adapter/api3/API3Oracle.sol [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    string public constant name = "API3Oracle";
+    /// @notice The minimum permitted value for `maxStaleness`.
+    uint256 internal constant MAX_STALENESS_LOWER_BOUND = 1 minutes;
+    /// @notice The maximum permitted value for `maxStaleness`.
+    uint256 internal constant MAX_STALENESS_UPPER_BOUND = 72 hours;
+    /// @notice The address of the base asset corresponding to the feed.
+    address public immutable base;
+    /// @notice The address of the quote asset corresponding to the feed.
+    address public immutable quote;
+    /// @notice The address of the Chainlink price feed.
```
Comment (totomanov) [inline src/adapter/api3/API3Oracle.sol:19]: nit: rename Chainlink -> API3

File: src/adapter/api3/API3Oracle.sol [hunk 2]
```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.0;
+
+import {BaseAdapter, Errors, IPriceOracle} from "src/adapter/BaseAdapter.sol";
+import {IDapiProxy} from "src/adapter/api3/IDapiProxy.sol";
+import {ScaleUtils, Scale} from "src/lib/ScaleUtils.sol";
+
+contract API3Oracle is BaseAdapter {
```
Comment (totomanov) [inline src/adapter/api3/API3Oracle.sol:8]: nit: add contract-level natspec

File: src/adapter/api3/API3Oracle.sol [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    /// @notice The maximum allowed age of the price.
+    /// @dev Reverts if block.timestamp - updatedAt > maxStaleness.
+    uint256 public immutable maxStaleness;
+    /// @notice The scale factors used for decimal conversions.
+    Scale internal immutable scale;
+
+    /// @notice Deploy a API3Oracle.
+    /// @param _base The address of the base asset corresponding to the feed.
+    /// @param _quote The address of the quote asset corresponding to the feed.
+    /// @param _feed The address of the Chainlink price feed.
```
Comment (totomanov) [inline src/adapter/api3/API3Oracle.sol:31]: nit: rename Chainlink -> API3

File: src/adapter/api3/IDapiProxy.sol [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+/// @title IDapiProxy
+/// @author api3dao (https://github.com/api3dao/contracts/blob/e9487f14db13edf60a10ba3ece08c9a6c0e7f9a9/contracts/api3-server-v1/proxies/interfaces/IProxy.sol)
+/// @notice Partial interface for API3 Data Feeds.
+interface IDapiProxy {
+    /// @notice Reads the dAPI that this proxy maps to
+    function read() external view returns (int224 value, uint32 timestamp);
+
+    /// @notice Api3ServerV1 address
+    function api3ServerV1() external view returns (address);
```
Comment (totomanov) [inline src/adapter/api3/IDapiProxy.sol:12]: looks like this is not needed, so we can remove it

Comment (wiasliaw) [reply src/adapter/api3/IDapiProxy.sol]: Resolved.

File: src/adapter/api3/API3Oracle.sol (hunk 3 above)
Comment (wiasliaw) [reply src/adapter/api3/API3Oracle.sol]: Resolved.

File: src/adapter/api3/API3Oracle.sol (hunk 2 above)
Comment (wiasliaw) [reply src/adapter/api3/API3Oracle.sol:14]: Resolved.

File: src/adapter/api3/API3Oracle.sol (hunk 1 above)
Comment (wiasliaw) [reply src/adapter/api3/API3Oracle.sol]: Resolved.

**Comment by totomanov** [comment]
Thanks, merging. Feel free to submit tests in a separate PR.

**Pull request merged**